---
title: "Functional Analysis"
subtitle: "From Raw Counts to Differentially Expressed Genes"
format:
  revealjs: 
    slide-number: true
    chalkboard: 
      buttons: false
    preview-links: auto
    logo: img/eda.svg
    css: styles.css
    incremental: true
---

## Preprocessing {.smaller}

### Why?
- Sequencing depth varies across libraries
- High dynamic range of expression
- High number of genes vs low number of samples

### What?
- Filtering of low abundance genes
- Normalization 
  + Library size
  + Gene length

![](img/data-cleaning.svg){.absolute top="250" right="70" height="250"}

## Filtering
<br>

__Simple__: 
Keep genes with more than *X* counts

```r
# Keep genes/rows that have > 5 total counts 
se <- se[rowSums(assay(se, "counts")) > 5, ]
```
<br>

__Advanced__: 
Keep genes with more than *X* counts in at least *Y* samples

```r
# Keep genes/rows that have > 3 counts in at least 3 sample
se[rowSums(assay(se, "counts") > 3) >= 3, ]
```

## Filtering - Considerations

- __Easier Interpretation__: Not or lowly expressed genes are unlikely to be biological meaningful
- __More Power__: Filtering reduces the number of statistical tests (multiple testing)
- __Better Model__: More roboust estimation of mean-variance relationship

::: {.fragment .highlight-red}
Try different approaches and check the remaining gene number and type.
:::

## Normalization - Gene Length {.smaller}
<br>

::: {.fragment .highlight-red}
*Making counts comparable across features (genes)*
:::
<br>

![](img/normalization_methods_length.png){height="500"}

## Normalization - Sequencing Depth {.smaller}
<br>

::: {.fragment .highlight-red}
*Making counts comparable across samples*
:::
<br>
![](img/normalization_methods_depth.png){height="500"}


## Normalization - Approaches {.smaller}
<br>

::: columns
::: {.column width="55%"}
- by library size (using gene length and sequencing depth)
- by distribution (group genes with similar expression)
- by testing (identify non-DE genes for normalization)
- using controls (e.g. spike-ins, house-keepig genes)
- variance-stabilizing transformation (stable variance across different means)
:::
::: {.column width="45%"}
![](img/normalization_methods_composition.png) 
:::
:::


## {.smaller}

| Normalization method | Description | Accounted factors | Recommendations for use |
| :----: | :----: | :----: | :----: |
| **CPM** (counts per million) | counts scaled by total number of reads | sequencing depth | gene count comparisons between replicates; **NOT for within sample or DE analysis**  |
| **TPM** (transcripts per kilobase million) | counts per length of transcript (kb) per million reads mapped | sequencing depth and gene length | gene count comparisons within a sample or sample group; **NOT for DE analysis** |
| **RPKM/FPKM** (reads/fragments per kilobase of exon per million reads/fragments mapped) | similar to TPM | sequencing depth and gene length | gene count comparisons within a sample; **NOT for between sample or DE analysis** |

::: footer
Table from [hbctraing:Intro-to-DGE](https://hbctraining.github.io/Intro-to-DGE)
:::

## {.smaller}

| Normalization method | Description | Accounted factors | Recommendations for use |
| :----: | :----: | :----: | :----: |
| DESeq2's **median of ratios** [[1](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-10-r106)] | counts divided by sample-specific size factors determined by median ratio of gene counts relative to geometric mean per gene | sequencing depth and RNA composition | gene count comparisons between samples and for **DE analysis**; **NOT for within sample comparisons** |
| EdgeR's **trimmed mean of M values (TMM)** [[2](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-3-r25)] | uses a weighted trimmed mean of the log expression ratios between samples | sequencing depth and RNA composition | gene count comparisons between samples and for **DE analysis**; **NOT for within sample comparisons** |

::: footer
Table from [hbctraing:Intro-to-DGE](https://hbctraining.github.io/Intro-to-DGE)
:::


## Recommendations

- Under the hood most tools use a mix of different normalisations (e.g. `DESeq2::estimateSizeFactors()`)
- Standard tools for DGE (DESeq2, edgeR, Limma etc) use __raw counts__
- For visualisation (PCA, clustering, heatmaps etc), use VST or RLOG
- For own analysis with gene length correction, use TPM (maybe geTMM?)


## Exploratory Data Analysis
<br>

::: {.fragment .highlight-red}
*Inspect your data and identify and try to understand general pattern*
:::
<br>

### Examples
- Clustering
- Principal Component Ananlysis

![](img/data-viz.svg){.absolute top="300" right="80" height="350"}

## Principal Component Analysis
<br>

::: columns
::: {.column width="55%"}
- Reduce complexity and understand main driver of variance
- Map data into low dimensional space (*linear dimensionality reduction*)
- Interpretable 

:::
::: {.column width="45%"}
![](img/red-dim.png) 

:::
:::


## Principal Component Analysis
<br>
First dimensions represent main axis of variance in data

![](img/pca-contrib.png) 

::: footer
Czarnewski, P., Parigi, S.M., Sorini, C. et al. Conserved transcriptomic profile between mouse and human colitis allows unsupervised patient stratification. Nat Commun 10, 2892 (2019). https://doi.org/10.1038/s41467-019-10769-x
:::

## Principal Component Analysis
<br>

![](img/shakira.svg) 

::: footer
from https://glowingpython.blogspot.com/2011/07/pca-and-image-compression-with-numpy.html
:::

## Clustering {.smaller}

- Classify/group samples -> understand sample similarities
- Understand pattern
- Identify outlier
- Try different clustering methods and check cluster roboustness

![](img/clustering.png){height="480"}

## Differential Expression{.smaller}

<br>

::: {.fragment .highlight-red}
*Understanding systematic changes between experimental groups or conditions*
:::
<br>

::: columns
::: {.column width="45%"}
- Identify genes with differential expression pattern using *between*- and *within-condition* variability
- Requires __biological replicates__ (multiple sample from each condition)
- Multiple R packages exist specific for DE analysis (e.g. [edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html), [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html))

:::
::: {.column width="55%"}
![](img/de_theory.png)

:::
:::

::: footer
Image from [hbctraing:Intro-to-DGE](https://hbctraining.github.io/Intro-to-DGE)
:::

## Differential Expression Analysis

### RNAseq Data Challenges:
- Limited number of replicates per group
- Mean-variance dependency
- Skewed count distribution

![](img/deseq_counts_overview.png)

::: footer
Image from [hbctraing:Intro-to-DGE](https://hbctraining.github.io/Intro-to-DGE)
:::

## DESeq2 {.smaller}

![](img/DESeq2.png){.absolute top="40" right="80" height="100"}

<br>

::: columns
::: {.column width="45%"}
#### Statistical modeling
- model RNA-seq counts as *negative binomial distribution* 
- borrow information from genes with similar expression pattern to estimate *gene-wise dispersion*
- Fit a *Generalized linear model* for each gene and run Wald test

:::
::: {.column width="55%"}
![](img/deseq_dispersion2.png){top="170"}

:::
:::

::: footer
Image from [hbctraing:Intro-to-DGE](https://hbctraining.github.io/Intro-to-DGE)
:::

## The Design Matrix and Contrasts {.smaller}

 <br>

#### Design matrix

::: {.nonincremental}
- Describes the model coefficients
- Include variable of interest and all variables to account for
:::

```r
designFormula = ~ sex + time
```

 <br>

#### Contrasts

::: {.nonincremental}
- Describes the comparison of interest (reference variable and levels to compare)
:::

```r
results(dds, contrast = c("time", "Day8", "Day0"))
```

## DE Results

<br>

::: columns
::: {.column width="45%"}

#### Heatmap
![](img/heatmap.png)

:::
::: {.column width="55%"}
#### Volcano plot
<br>
<br>
![](img/volcano.png)

:::
:::