---
title: "Introduction to RNA-Seq"
subtitle: "From Experiment to Count Table"
format:
  revealjs: 
    slide-number: true
    chalkboard: 
      buttons: false
    preview-links: auto
    css: styles.css
---

## The Central Dogma of Molecular Biology

![](img/centralDogma.png){.absolute top="170" left="30" width="820" height="500"}

## The Reality is more complex ...
::: incremental
-   RNA is not only a messenger but may have other roles
-   Most of these elements interacts and regulates one-another
-   Additional layer of complexity through alternative splicing
:::

. . .

::: columns
::: {.column width="50%"}
![](img/centralDogma.png){.absolute top="320" left="30" width="450" height="300"}
:::
::: {.column width="50%"}
::: {.fragment .highlight-red}
mRNA levels → a proxy of a proxy of cell physiology
:::
:::
:::

## What are we sequencing?
#### Genomics
- Whole genome/exome sequencing (WGS/WES)
- Variant calling (SNPs, CNVs, structural variations)

#### Epigenomics
- Bisulphite sequencing : DNA methylation
- ATAC-Seq : chromatine opening
- ChIP-seq : TF binding sites

## What are we sequencing? {.scrollable}
#### Transcriptomics
- Total RNA
- Poly-A tail selection : focus on mRNA
- Ribo depletion: mRNA + ncRNA
- 5'/3' RACE seq : isoform characterization for one gene
- single cell RNAseq
- Long read RNA sequencing
- Spatial Transcriptomics

many more ...

## Single cell versus bulk RNA-Seq {.smaller}

![](img/sc_vs_bulk.jpg){.border}

::: footer
Zhu et al. J Transl Med 2024
:::

## Why bulk RNA-seq still matters in the era of single cell RNA-seq

. . .

<br/> Cost-effectiveness and Scalability

::: incremental
- Bulk RNA-seq is orders of magnitude cheaper
- Enables large cohorts (hundreds–thousands of samples)
- Essential for population-level or clinical studies
:::

## Why bulk RNA-seq still matters in the era of single cell RNA-seq

<br/> Higher Data Quality per Gene

::: incremental
- Deeper sequencing per transcript → better quantification
- Less technical noise, no dropouts
- More reliable differential expression results
:::

## Why bulk RNA-seq still matters in the era of single cell RNA-seq

<br/> Simplicity and Robustness

::: incremental
- Well-established workflows and QC pipelines
- Works with degraded/frozen tissues (recent scRNAseq techniques as well)
- Relative high reproducibility across labs
:::

## Why bulk RNA-seq still matters in the era of single cell RNA-seq

<br/> Biological Interpretability

::: incremental
- Average expression is often the relevant measure for tissue-level processes
- Examples: Treatment effects on cell line, Developmental stages, Systemic disease states
:::

## RNAseq data generation - planning your experiment 

::: columns
::: {.column width="40%"}

::: incremental
- How many replicates?
- Technical: same RNA extract
- Biological: same biological condition
- Sequencing depth?
- RNA purification?
:::
:::

::: {.column width="50%"}
![](img/lab_workflow.png){.absolute top="100" left="500" width="480" height="600"}
:::
:::

## Biological Replicates

::: columns
::: {.column width="50%"}
![](img/biological_rep_FDR.png)
:::

::: {.column width="50%"}
![](img/biological_rep_TPR.png)
:::
:::

::: footer
Soneson, C., et al. BMC Bioinformatics 14, 91 (2013)
:::

## Sequencing Depth

::: columns
::: {.column width="40%"}

::: incremental

- For DGE analysis: usually aim for 30-40 million reads
- Rare events require a much higher sequencing depths
- often hard to know in advance
:::
:::

::: {.column width="8%"}
:::

::: {.column width="50%"}
![](img/seq_depth.png)
:::
:::

## Data Limitations: Sequencing Biases

::: incremental

- Fragment length: longer fragments are more likely to be identified
- Positional bias: 3' ends are more likely to be sequenced
- Fragment sequence bias: Sequence complexity and GC content influence primer binding
:::

![](img/biases.png){.absolute top="390" left="30" width="1200" height="270"}

## Basic Data analysis Pipeline
![](img/analysis_wf.png)

## Input data 
. . . 

Raw data - FASTQ (.fq/.fastq)

::: incremental
- contains raw sequencing reads plus quality scores
:::

. . .

Reference - FASTA (.fa/.fasta)

::: incremental
- contains nucleotide sequences
:::

. . .

Annotation file - GTF/GFF (.gff/.gtf)

::: incremental
- contains genomic coordinates and features (e.g. genes, exon, ..)
:::

## Preprocessing
![](img/analysis_wf_sub1.png)

## Quality Control
. . .

Aim: Identify potential quality issues before analysis

. . .

Tools: FastQC, MultiQC

![](img/fastqc_summary.png)

## Quality Control

::: columns
::: {.column width="50%"}
Key Metrics

- Per-base quality scores
- GC content distribution
- Sequence length distribution 
- Overrepresented sequences
:::
::: {.column width="50%"}
![](img/fastqc_pbq.png)
:::
:::

## Trimming
. . .

Aim: Removal of adapter sequences and low-quality regions

. . .

Tools: Trimmomatic, Cutadaptt, fastp

![](img/trimming_QC.png)

## Trimming

Caution: Aggressive trimming (high quality threshold, low length filtering) can have negative impacts on expression quantification (William et al. BMC bioinfo, 2016)

. . . 

When should we trim?

::: incremental

- Variant calling: highly recommended
- Expression quantification: no or light trimming
- Transcriptome asssembly: highly recommended
:::


## Read Alignment and Quantification
![](img/analysis_wf.png)

## Read Alignment and Quantification

Aim: Map each read base-by-base to reference genome/transcriptome and count assigned reads

. . .

Mapping Tools: STAR, HISAT2, Bowtie2, BWA, ... (many more)

. . .

Quantification tools: featureCounts, HTSeq-count


## Choosing a reference: Genome vs Transcriptome
. . .

Genome reference:

::: incremental
- Whole genome including splicing junctions
- Use Case: Novel transcript discovery, variant calling
:::

. . .

Transcriptome reference:

::: incremental
- Set of known transcripts (exons only)
- Use Case: Fast quantification in well-annotated species
:::

## Pseudoalignment

::: incremental
- Reads are assigned to transcripts probabilitically
- Much faster; no full alignment
- Tools: Salmon, Kallisto
:::

![](img/analysis_wf_pseudo.png)


## Read Normalization
. . .

Aim: Adjust for sequencing depth and gene length

. . .

Methods:

::: incremental
- TPM: Within-sample normalization
- FPKM/RPKM: less used today
:::

. . .

Important: Some tools for DE analysis (DESeq2, edgeR) have their own normalization implemented

## Downstream Analysis
![](img/analysis_wf_final.png)